algorithms/kernel/svm/svm_train_common.h

```diff
@@ -0,0 +1,171 @@
+    {
+        return (y > 0 && alpha > 0) || (y < 0 && alpha < C);
+    }
+
+    DAAL_FORCEINLINE static algorithmFPType WSSi(size_t nActiveVectors, const algorithmFPType * grad, const char * I, int & Bi);
+
+    DAAL_FORCEINLINE static void WSSjLocal(const size_t jStart, const size_t jEnd, const algorithmFPType * KiBlock,
+                                           const algorithmFPType * kernelDiag, const algorithmFPType * grad, const char * I,
+                                           const algorithmFPType GMin, const algorithmFPType Kii, const algorithmFPType tau, int & Bj,
+                                           algorithmFPType & GMax, algorithmFPType & GMax2, algorithmFPType & delta);
+
+private:
+    DAAL_FORCEINLINE static void WSSjLocalBaseline(const size_t jStart, const size_t jEnd, const algorithmFPType * KiBlock,
+                                                   const algorithmFPType * kernelDiag, const algorithmFPType * grad, const char * I,
+                                                   const algorithmFPType GMin, const algorithmFPType Kii, const algorithmFPType tau, int & Bj,
+                                                   algorithmFPType & GMax, algorithmFPType & GMax2, algorithmFPType & delta);
+};
+
+template <typename algorithmFPType, CpuType cpu>
+class SubDataTaskBase
+{
+public:
+    DAAL_NEW_DELETE();
+    virtual ~SubDataTaskBase() {}
+
+    virtual services::Status copyDataByIndices(const uint32_t * wsIndices, const NumericTablePtr & xTable) = 0;
+
+    NumericTablePtr getTableData() const { return _dataTable; }
+
+protected:
+    SubDataTaskBase(const size_t nSubsetVectors, const size_t dataSize) : _nSubsetVectors(nSubsetVectors), _data(dataSize) {}
+    SubDataTaskBase(const size_t nSubsetVectors) : _nSubsetVectors(nSubsetVectors) {}
+
+    bool isValid() const { return _data.get(); }
+
+protected:
+    size_t _nSubsetVectors;
+    TArray<algorithmFPType, cpu> _data;
+    NumericTablePtr _dataTable;
+};
+
+template <typename algorithmFPType, CpuType cpu>
+class SubDataTaskCSR : public SubDataTaskBase<algorithmFPType, cpu>
+{
+public:
+    using super = SubDataTaskBase<algorithmFPType, cpu>;
+    static SubDataTaskCSR * create(const NumericTablePtr & xTable, const size_t nSubsetVectors)
+    {
+        auto val = new SubDataTaskCSR(xTable, nSubsetVectors);
+        if (val && val->isValid()) return val;
+        delete val;
+        val = nullptr;
+        return nullptr;
+    }
+
+private:
+    bool isValid() const { return super::isValid() && _colIndices.get() && this->_dataTable.get(); }
+
+    SubDataTaskCSR(const NumericTablePtr & xTable, const size_t nSubsetVectors) : super(nSubsetVectors)
+    {
+        const size_t p                  = xTable->getNumberOfColumns();
+        const size_t nRows              = xTable->getNumberOfRows();
+        CSRNumericTableIface * csrIface = dynamic_cast<CSRNumericTableIface *>(const_cast<NumericTable *>(xTable.get()));
```

Comment:
should this pointer be checked on 'nullptr' ?

Reply:
Fixed